packages/runtime/src/runtime/screen.tsx

```diff
@@ -91,34 +93,76 @@ export const EnsembleScreen: React.FC<EnsembleScreenProps> = ({
     };
   }, [screen.customWidgets]);
 
+  const AppScreen = useCallback(() => {
+    if (screen.apis?.length !== apis.length) {
+      return null;
+    }
+
+    return (
+      <ScreenContextProvider
+        context={{ inputs: mergedInputs }}
+        key={pathname}
+        screen={{ ...screen, apis }}
+      >
+        <ModalWrapper>
+          <OnLoadAction action={screen.onLoad} context={{ ...mergedInputs }}>
+            <EnsembleHeader header={screen.header} />
+            <EnsembleBody body={screen.body} styles={screen.styles} />
+          </OnLoadAction>
+          {screen.menu ? (
+            <EnsembleMenu
+              menu={{ ...screen.menu }}
+              renderOutlet={false}
+              type={screen.menu.type}
+            />
+          ) : null}
+          <EnsembleFooter footer={screen.footer} />
+        </ModalWrapper>
+      </ScreenContextProvider>
+    );
+  }, [apis]);
+
   if (!isInitialized) {
     return null;
   }
 
   return (
-    <ScreenContextProvider
-      context={{ inputs: mergedInputs }}
-      key={pathname}
-      screen={screen}
-    >
-      <ModalWrapper>
-        <OnLoadAction action={screen.onLoad} context={{ ...mergedInputs }}>
-          <EnsembleHeader header={screen.header} />
-          <EnsembleBody body={screen.body} styles={screen.styles} />
-        </OnLoadAction>
-        {screen.menu ? (
-          <EnsembleMenu
-            menu={{ ...screen.menu }}
-            renderOutlet={false}
-            type={screen.menu.type}
-          />
-        ) : null}
-        <EnsembleFooter footer={screen.footer} />
-      </ModalWrapper>
-    </ScreenContextProvider>
+    <>
+      {screen.apis?.map((api, index) => (
+        <OnEvaluateApi
+          api={api}
+          currentIndex={index}
+          key={`${api.name}_${index}`}
+          setApis={setApis}
+        />
+      ))}
+      <AppScreen />
+    </>
   );
 };
 
+const OnEvaluateApi = ({
+  api,
+  currentIndex,
+  setApis,
+}: {
+  api: EnsembleAPIModel;
+  currentIndex: number;
+  setApis: React.Dispatch<React.SetStateAction<EnsembleAPIModel[]>>;
+}): null => {
+  const onResponse = useEnsembleAction(api.onResponse as EnsembleAction);
+  const onError = useEnsembleAction(api.onError as EnsembleAction);
```

Comment:
This needs to happen inside the ScreenProvider/ModalWrapper.

Reply:
I think this is the best place to convert that into callback, because we need to this in runtime only. other thing is to after converting them into callback, we need to store them in screenAtom, thus we can consume that into framework by accessing from screenAtom by screenContext.

Reply:
I don't think you are understanding me here. This evaluation needs to happen inside of the screen and modal context otherwise the other actions won't work